src/clientHandler.ts

```diff
@@ -0,0 +1,252 @@
+			if (this.GetClient()?.client.needsStart()) {
+				console.log(`No need to start another client for ${folders}`)
+				return disposables;
+			}
+
+			console.log('Starting client');
+
+			let tfClient = this.newTerraformClient();
+			tfClient.client.onReady().then(async () => {
+				this.reporter.sendTelemetryEvent('startClient');
+				const multiFoldersSupported = tfClient.client.initializeResult.capabilities.workspace?.workspaceFolders?.supported;
+				console.log(`Multi-folder support: ${multiFoldersSupported}`);
+
+				if (!multiFoldersSupported) {
+					// restart is needed to launch folder-focused instances
+					console.log('Restarting clients as folder-focused');
+					await this.StopClients(folders);
+					this.supportsMultiFolders = false;
+					this.StartClients(folders);
+				}
+			});
+
+			disposables.push(tfClient.client.start());
+			clients.set("", tfClient);
+
+			return disposables;
+		}
+
+		if (folders && folders.length > 0) {
+			for (const folder of folders) {
+				if (!clients.has(folder)) {
+					console.log(`Starting client for ${folder}`);
+					let folderClient = this.newTerraformClient(folder);
+					folderClient.client.onReady().then(() => {
+						this.reporter.sendTelemetryEvent('startClient');
+					});
+
+					disposables.push(folderClient.client.start());
+					clients.set(folder, folderClient);
+				} else {
+					console.log(`Client for folder: ${folder} already started`);
+				}
+			}
+		}
+		return disposables;
+	};
+
+	private newTerraformClient(location?: string): TerraformLanguageClient {
+		const cmd = this.pathToBinary;
+		const binaryName = cmd.split('/').pop();
+
+		const serverArgs: string[] = config('terraform').get('languageServer.args');
+		const experimentalFeatures = config('terraform-ls').get('experimentalFeatures');
+
+		let channelName = `${binaryName}`;
+		let id = `languageServer`
+		let name = `Language Server`
+		let wsFolder: vscode.WorkspaceFolder;
+		let rootModulePaths: string[];
+		let excludeModulePaths: string[];
+		let documentSelector: DocumentSelector;
+		let outputChannel: vscode.OutputChannel;
+		if (location) {
+			channelName = `${binaryName}: ${location}`;
+			id = `languageServer/${location}`
+			name = `Language Server: ${location}`
+			wsFolder = getWorkspaceFolder(location);
+			documentSelector = [
+				{ scheme: 'file', language: 'terraform', pattern: `${wsFolder.uri.fsPath}/**/*` },
+				{ scheme: 'file', language: 'terraform-vars', pattern: `${wsFolder.uri.fsPath}/**/*` }
+			]
+			rootModulePaths = config('terraform-ls', wsFolder).get('rootModules');
+			excludeModulePaths = config('terraform-ls', wsFolder).get('excludeRootModules');
+			outputChannel = vscode.window.createOutputChannel(channelName);
+			outputChannel.appendLine(`Launching language server: ${cmd} ${serverArgs.join(' ')} for folder: ${location}`);
+		} else {
+			documentSelector = [
+				{ scheme: 'file', language: 'terraform' },
+				{ scheme: 'file', language: 'terraform-vars' }
+			]
```

Comment:
This is the key part where we launch the client with appropriate selector depending on how it's being launched.

Reply:
Looks good 👍